numpy/core/src/multiarray/descriptor.c

```diff
@@ -241,6 +241,10 @@ is_datetime_typestr(char const *type, Py_ssize_t len)
 static PyArray_Descr *
 _convert_from_tuple(PyObject *obj, int align)
 {
+    if (Py_EnterRecursiveCall(
+            " while trying to create self-referential dtypes" ) != 0) {
+        return NULL;
+    }
```

Comment:
Instead of putting this code in four places, can you put it in the function that calls `_convert_from_tuple`?

Reply:
(which I think is `_convert_from_any`)

Reply:
Yeah you are right Eric, but seems like `_tuple`, `_list`, `_dict`, etc also calls `_convert_from_any`, which in turns calls it(the issue basically), let me refactor a bit, will get back.

Reply:
One way to think about it is to draw out the callgraph, and make sure you cut every loop with a recursive call check. What you have here works fine (and is what I originally suggested), but it may be easier to handle it you cut the loop somewhere else.

The advantage of what you have here is that you can tailor the message to each function.

Reply:
Hey @eric-wieser , I took the best of both methods and added sperate interceptors. Let me know if this idea is ok, thanks.